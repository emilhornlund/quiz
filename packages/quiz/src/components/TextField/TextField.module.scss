@import '../../styles/colors.scss';
@import '../../styles/helpers.scss';

$checkbox-height-ratio: 0.7;

$mobile-checkbox-normal-size: calc($mobile-control-normal-height * $checkbox-height-ratio);
$tablet-checkbox-normal-size: calc($tablet-control-normal-height * $checkbox-height-ratio);
$desktop-checkbox-normal-size: calc($desktop-control-normal-height * $checkbox-height-ratio);

$mobile-checkbox-small-size: calc($mobile-control-small-height * $checkbox-height-ratio);
$tablet-checkbox-small-size: calc($tablet-control-small-height * $checkbox-height-ratio);
$desktop-checkbox-small-size: calc($desktop-control-small-height * $checkbox-height-ratio);

$mobile-checkbox-normal-font-size: $mobile-control-normal-font-size;
$tablet-checkbox-normal-font-size: $tablet-control-normal-font-size;
$desktop-checkbox-normal-font-size: $desktop-control-normal-font-size;

$mobile-checkbox-small-font-size: calc($mobile-checkbox-normal-font-size * 0.8);
$tablet-checkbox-small-font-size: calc($tablet-checkbox-normal-font-size * 0.8);
$desktop-checkbox-small-font-size: calc($desktop-checkbox-normal-font-size * 0.8);

.main {
  position: relative;
  display: flex;
  flex-direction: row;
  flex-grow: 0;
  flex-shrink: 1;
  align-items: center;
  outline: $white solid $control-outside-border-width;
  outline-offset: $control-outside-border-width;
  border: $control-outside-border-width solid transparent;
  margin: calc($control-outside-border-width * 2);
  background-color: $white;
  overflow: hidden;

  &.disabled {
    background-color: $gray-1;

    .textfield {
      color: $gray-2;
    }
  }

  &.error {
    outline-color: $red-2;
  }

  .textfield {
    display: flex;
    flex-direction: column;
    flex: 1;
    appearance: none;
    outline: none;
    border: none;
    background-color: transparent;
    color: $gray-4;

    @include mobile {
      @include fontSize($mobile-control-normal-font-size);
    }

    @include tablet {
      @include fontSize($tablet-control-normal-font-size);
    }

    @include desktop {
      @include fontSize($desktop-control-normal-font-size);
    }
  }

  .checkboxLabel {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    background-color: transparent;
    border: $control-outside-border-width solid $green-2;
    border-radius: 50%;
    transition: background-color 0.3s ease, border-color 0.3s ease;

    input[type=checkbox] {
      display: none;
    }

    .checkboxIcon {
      color: $white;
    }

    &:has(> input:checked) {
      background-color: $green-1;
    }

    @include mobile {
      width: $mobile-checkbox-normal-size;
      height: $mobile-checkbox-normal-size;

      @include fontSize($tablet-checkbox-normal-font-size);
    }

    @include tablet {
      width: $tablet-checkbox-normal-size;
      height: $tablet-checkbox-normal-size;

      @include fontSize($tablet-checkbox-normal-font-size);
    }

    @include desktop {
      width: $desktop-checkbox-normal-size;
      height: $desktop-checkbox-normal-size;

      @include fontSize($desktop-checkbox-normal-font-size);
    }
  }

  @include mobile {
    height: $mobile-control-normal-height;
    padding: 0 $mobile-control-normal-layout-gutter;
    border-radius: $mobile-control-normal-border-radius;

    &.small {
      height: $mobile-control-small-height;
      padding: 0 $mobile-control-small-layout-gutter;
      border-radius: $mobile-control-small-border-radius;

      .textfield {
        @include fontSize($mobile-control-small-font-size);
      }

      .checkboxLabel {
        width: $mobile-checkbox-small-size;
        height: $mobile-checkbox-small-size;

        @include fontSize($mobile-checkbox-small-font-size);
      }
    }
  }

  @include tablet {
    height: $tablet-control-normal-height;
    padding: 0 $tablet-control-normal-layout-gutter;
    border-radius: $tablet-control-normal-border-radius;

    &.small {
      height: $tablet-control-small-height;
      padding: 0 $tablet-control-small-layout-gutter;
      border-radius: $tablet-control-small-border-radius;

      .textfield {
        @include fontSize($tablet-control-small-font-size);
      }

      .checkboxLabel {
        width: $tablet-checkbox-small-size;
        height: $tablet-checkbox-small-size;

        @include fontSize($tablet-checkbox-small-font-size);
      }
    }
  }

  @include desktop {
    height: $desktop-control-normal-height;
    padding: 0 $desktop-control-normal-layout-gutter;
    border-radius: $desktop-control-normal-border-radius;

    &.small {
      height: $desktop-control-small-height;
      padding: 0 $desktop-control-small-layout-gutter;
      border-radius: $desktop-control-small-border-radius;

      .textfield {
        @include fontSize($desktop-control-small-font-size);
      }

      .checkboxLabel {
        width: $desktop-checkbox-small-size;
        height: $desktop-checkbox-small-size;

        @include fontSize($desktop-checkbox-small-font-size);
      }
    }
  }
}