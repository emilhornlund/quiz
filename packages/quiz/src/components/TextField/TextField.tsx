import React, { ChangeEvent, useEffect, useState } from 'react'

import { classNames, isValidNumber } from '../../utils/helpers.ts'

import styles from './TextField.module.scss'

export interface TextFieldProps {
  id: string
  name?: string
  type: 'text' | 'number'
  size?: 'normal' | 'small'
  placeholder?: string
  value?: string | number
  min?: number
  max?: number
  regex?: RegExp
  required?: boolean
  disabled?: boolean
  onChange?: (value?: string | number) => void
  onValid?: (valid: boolean) => void
}

const TextField: React.FC<TextFieldProps> = ({
  id,
  name,
  type,
  size = 'normal',
  placeholder,
  value,
  min,
  max,
  regex,
  required,
  disabled,
  onChange,
  onValid,
}) => {
  const [internalValue, setInternalValue] = useState<string | number>(
    value ?? '',
  )

  useEffect(() => {
    setInternalValue(value ?? '')
  }, [value])

  const [valid, setValid] = useState<boolean>(false)
  const [hasFocus, setHasFocus] = useState<boolean>(false)

  useEffect(() => {
    let tmpValid = true
    if (type === 'number') {
      if (
        typeof internalValue !== 'number' ||
        !isValidNumber(internalValue, min, max)
      ) {
        tmpValid = false
      }
    } else if (type === 'text') {
      if (typeof internalValue !== 'string') {
        tmpValid = false
      } else if (required && !internalValue) {
        tmpValid = false
      } else if (internalValue && regex && !regex.test(internalValue)) {
        tmpValid = false
      }
    }
    onValid?.(tmpValid)
    setValid(tmpValid)
  }, [internalValue, type, min, max, regex, required, onValid])

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    let newValue: string | number = event.target.value
    if (type === 'number') {
      const parsedValue = parseInt(newValue, 10)
      newValue = isNaN(parsedValue) ? '' : parsedValue
    }
    setInternalValue(newValue)
    onChange?.(newValue)
  }

  return (
    <div
      className={classNames(
        styles.main,
        size === 'small' ? styles.small : undefined,
        !valid && hasFocus ? styles.error : undefined,
      )}>
      <input
        id={id}
        name={name ?? id}
        type={type}
        value={internalValue}
        min={min}
        max={max}
        placeholder={placeholder}
        disabled={disabled}
        onChange={handleChange}
        onFocus={() => setHasFocus(true)}
        onBlur={() => setHasFocus(false)}
        data-testid={`test-${id}-textfield`}
      />
    </div>
  )
}

export default TextField
