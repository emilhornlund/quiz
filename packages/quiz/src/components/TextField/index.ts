export type { TextFieldProps } from './TextField.tsx'
export { default } from './TextField.tsx'
