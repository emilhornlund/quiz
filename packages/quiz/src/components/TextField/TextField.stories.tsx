import type { Meta, StoryObj } from '@storybook/react'

import TextField from './TextField.tsx'

const meta = {
  component: TextField,
  tags: ['autodocs'],
} satisfies Meta<typeof TextField>

export default meta
type Story = StoryObj<typeof meta>

export const Text = {
  args: {
    id: 'my-text-field',
    type: 'text',
    placeholder: 'Placeholder',
  },
} satisfies Story

export const Number = {
  args: {
    id: 'my-text-field',
    type: 'number',
    placeholder: 'Placeholder',
    min: 0,
    max: 100,
  },
} satisfies Story

export const Disabled = {
  args: {
    id: 'my-text-field',
    type: 'text',
    placeholder: 'Placeholder',
    disabled: true,
  },
} satisfies Story

export const Small = {
  args: {
    id: 'my-small-text-field',
    type: 'text',
    size: 'small',
    placeholder: 'Placeholder',
  },
} satisfies Story
