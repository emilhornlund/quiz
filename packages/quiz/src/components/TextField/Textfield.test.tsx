import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { describe, expect, it, vi } from 'vitest'

import TextField from './TextField.tsx'

describe('TextField', () => {
  it('should render a TextField with type text', async () => {
    const { container } = render(<TextField id="my-textfield" type="text" />)

    const input = screen.getByRole('textbox')
    fireEvent.focus(input)
    fireEvent.change(input, { target: { value: 'some value' } })

    expect(container).toMatchSnapshot()
  })

  it('should render a TextField with type number', async () => {
    const { container } = render(<TextField id="my-textfield" type="number" />)

    expect(container).toMatchSnapshot()
  })

  it('should receive onChange event when text changes', async () => {
    const onChange = vi.fn()

    const { container } = render(
      <TextField id="my-textfield" type="text" onChange={onChange} />,
    )

    const input = screen.getByTestId(
      'test-my-textfield-textfield',
    ) as HTMLInputElement

    fireEvent.focus(input)
    fireEvent.change(input, { target: { value: 'some value' } })

    expect(input.value).toBe('some value')

    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenCalledWith('some value')

    expect(container).toMatchSnapshot()
  })

  it('should receive onChange event when number changes', async () => {
    const onChange = vi.fn()

    const { container } = render(
      <TextField id="my-textfield" type="number" onChange={onChange} />,
    )

    const input = screen.getByTestId(
      'test-my-textfield-textfield',
    ) as HTMLInputElement

    fireEvent.focus(input)
    fireEvent.change(input, { target: { value: '1337' } })

    expect(input.value).toBe('1337')

    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenCalledWith(1337)

    expect(container).toMatchSnapshot()
  })

  it('should receive onChange event when number equal min changes', async () => {
    const onChange = vi.fn()

    const { container } = render(
      <TextField id="my-textfield" type="number" min={0} onChange={onChange} />,
    )

    const input = screen.getByTestId(
      'test-my-textfield-textfield',
    ) as HTMLInputElement

    fireEvent.focus(input)
    fireEvent.change(input, { target: { value: '0' } })

    expect(input.value).toBe('0')

    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenCalledWith(0)

    expect(container).toMatchSnapshot()
  })

  it('should receive onChange event when number equal max changes', async () => {
    const onChange = vi.fn()

    const { container } = render(
      <TextField
        id="my-textfield"
        type="number"
        max={100}
        onChange={onChange}
      />,
    )

    const input = screen.getByTestId(
      'test-my-textfield-textfield',
    ) as HTMLInputElement

    fireEvent.focus(input)
    fireEvent.change(input, { target: { value: '100' } })

    expect(input.value).toBe('100')

    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenCalledWith(100)

    expect(container).toMatchSnapshot()
  })

  it('should not receive onChange event when not a number changes', async () => {
    const onChange = vi.fn()

    const { container } = render(
      <TextField
        id="my-textfield"
        type="number"
        min={0}
        max={100}
        onChange={onChange}
      />,
    )

    const input = screen.getByTestId(
      'test-my-textfield-textfield',
    ) as HTMLInputElement

    fireEvent.focus(input)
    fireEvent.change(input, { target: { value: 'not a number' } })

    expect(input.value).toBe('')

    expect(onChange).not.toHaveBeenCalled()

    expect(container).toMatchSnapshot()
  })

  it('should not receive onChange event when number bellow min changes', async () => {
    const onChange = vi.fn()

    const { container } = render(
      <TextField
        id="my-textfield"
        type="number"
        min={0}
        max={100}
        onChange={onChange}
      />,
    )

    const input = screen.getByTestId(
      'test-my-textfield-textfield',
    ) as HTMLInputElement

    fireEvent.focus(input)
    fireEvent.change(input, { target: { value: '-1' } })

    expect(onChange).toHaveBeenCalled()

    expect(container).toMatchSnapshot()
  })

  it('should not receive onChange event when number above max changes', async () => {
    const onChange = vi.fn()

    const { container } = render(
      <TextField
        id="my-textfield"
        type="number"
        min={0}
        max={100}
        onChange={onChange}
      />,
    )

    const input = screen.getByTestId(
      'test-my-textfield-textfield',
    ) as HTMLInputElement

    fireEvent.focus(input)
    fireEvent.change(input, { target: { value: '101' } })

    expect(onChange).toHaveBeenCalled()

    expect(container).toMatchSnapshot()
  })
})
