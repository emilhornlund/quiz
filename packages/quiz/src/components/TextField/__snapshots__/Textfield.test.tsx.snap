// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextField > should not receive onChange event when not a number changes 1`] = `
<div>
  <div
    class="inputContainer"
  >
    <div
      class="textFieldInputContainer textFieldInputKindPrimary error"
    >
      <input
        class="textfield"
        data-testid="test-my-textfield-textfield"
        id="my-textfield"
        max="100"
        min="0"
        name="my-textfield"
        type="number"
        value=""
      />
    </div>
    <div
      class="errorContainer"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-triangle-exclamation"
        data-icon="triangle-exclamation"
        data-prefix="fas"
        role="img"
        viewBox="0 0 512 512"
      >
        <path
          d="M256 0c14.7 0 28.2 8.1 35.2 21l216 400c6.7 12.4 6.4 27.4-.8 39.5S486.1 480 472 480L40 480c-14.1 0-27.2-7.4-34.4-19.5s-7.5-27.1-.8-39.5l216-400c7-12.9 20.5-21 35.2-21zm0 352a32 32 0 1 0 0 64 32 32 0 1 0 0-64zm0-192c-18.2 0-32.7 15.5-31.4 33.7l7.4 104c.9 12.5 11.4 22.3 23.9 22.3 12.6 0 23-9.7 23.9-22.3l7.4-104c1.3-18.2-13.1-33.7-31.4-33.7z"
          fill="currentColor"
        />
      </svg>
       
      Must be a valid number
    </div>
  </div>
</div>
`;

exports[`TextField > should not receive onChange event when number above max changes 1`] = `
<div>
  <div
    class="inputContainer"
  >
    <div
      class="textFieldInputContainer textFieldInputKindPrimary error"
    >
      <input
        class="textfield"
        data-testid="test-my-textfield-textfield"
        id="my-textfield"
        max="100"
        min="0"
        name="my-textfield"
        type="number"
        value="101"
      />
    </div>
    <div
      class="errorContainer"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-triangle-exclamation"
        data-icon="triangle-exclamation"
        data-prefix="fas"
        role="img"
        viewBox="0 0 512 512"
      >
        <path
          d="M256 0c14.7 0 28.2 8.1 35.2 21l216 400c6.7 12.4 6.4 27.4-.8 39.5S486.1 480 472 480L40 480c-14.1 0-27.2-7.4-34.4-19.5s-7.5-27.1-.8-39.5l216-400c7-12.9 20.5-21 35.2-21zm0 352a32 32 0 1 0 0 64 32 32 0 1 0 0-64zm0-192c-18.2 0-32.7 15.5-31.4 33.7l7.4 104c.9 12.5 11.4 22.3 23.9 22.3 12.6 0 23-9.7 23.9-22.3l7.4-104c1.3-18.2-13.1-33.7-31.4-33.7z"
          fill="currentColor"
        />
      </svg>
       
      Cannot be greater than 100
    </div>
  </div>
</div>
`;

exports[`TextField > should not receive onChange event when number bellow min changes 1`] = `
<div>
  <div
    class="inputContainer"
  >
    <div
      class="textFieldInputContainer textFieldInputKindPrimary error"
    >
      <input
        class="textfield"
        data-testid="test-my-textfield-textfield"
        id="my-textfield"
        max="100"
        min="0"
        name="my-textfield"
        type="number"
        value="-1"
      />
    </div>
    <div
      class="errorContainer"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-triangle-exclamation"
        data-icon="triangle-exclamation"
        data-prefix="fas"
        role="img"
        viewBox="0 0 512 512"
      >
        <path
          d="M256 0c14.7 0 28.2 8.1 35.2 21l216 400c6.7 12.4 6.4 27.4-.8 39.5S486.1 480 472 480L40 480c-14.1 0-27.2-7.4-34.4-19.5s-7.5-27.1-.8-39.5l216-400c7-12.9 20.5-21 35.2-21zm0 352a32 32 0 1 0 0 64 32 32 0 1 0 0-64zm0-192c-18.2 0-32.7 15.5-31.4 33.7l7.4 104c.9 12.5 11.4 22.3 23.9 22.3 12.6 0 23-9.7 23.9-22.3l7.4-104c1.3-18.2-13.1-33.7-31.4-33.7z"
          fill="currentColor"
        />
      </svg>
       
      Cannot be less than 0
    </div>
  </div>
</div>
`;

exports[`TextField > should receive onChange event when number changes 1`] = `
<div>
  <div
    class="inputContainer"
  >
    <div
      class="textFieldInputContainer textFieldInputKindPrimary"
    >
      <input
        class="textfield"
        data-testid="test-my-textfield-textfield"
        id="my-textfield"
        name="my-textfield"
        type="number"
        value="1337"
      />
    </div>
  </div>
</div>
`;

exports[`TextField > should receive onChange event when number equal max changes 1`] = `
<div>
  <div
    class="inputContainer"
  >
    <div
      class="textFieldInputContainer textFieldInputKindPrimary"
    >
      <input
        class="textfield"
        data-testid="test-my-textfield-textfield"
        id="my-textfield"
        max="100"
        name="my-textfield"
        type="number"
        value="100"
      />
    </div>
  </div>
</div>
`;

exports[`TextField > should receive onChange event when number equal min changes 1`] = `
<div>
  <div
    class="inputContainer"
  >
    <div
      class="textFieldInputContainer textFieldInputKindPrimary"
    >
      <input
        class="textfield"
        data-testid="test-my-textfield-textfield"
        id="my-textfield"
        min="0"
        name="my-textfield"
        type="number"
        value="0"
      />
    </div>
  </div>
</div>
`;

exports[`TextField > should receive onChange event when text changes 1`] = `
<div>
  <div
    class="inputContainer"
  >
    <div
      class="textFieldInputContainer textFieldInputKindPrimary"
    >
      <input
        class="textfield"
        data-testid="test-my-textfield-textfield"
        id="my-textfield"
        name="my-textfield"
        type="text"
        value="some value"
      />
    </div>
  </div>
</div>
`;

exports[`TextField > should render a TextField with type number 1`] = `
<div>
  <div
    class="inputContainer"
  >
    <div
      class="textFieldInputContainer textFieldInputKindPrimary"
    >
      <input
        class="textfield"
        data-testid="test-my-textfield-textfield"
        id="my-textfield"
        name="my-textfield"
        type="number"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`TextField > should render a TextField with type text 1`] = `
<div>
  <div
    class="inputContainer"
  >
    <div
      class="textFieldInputContainer textFieldInputKindPrimary"
    >
      <input
        class="textfield"
        data-testid="test-my-textfield-textfield"
        id="my-textfield"
        name="my-textfield"
        type="text"
        value="some value"
      />
    </div>
  </div>
</div>
`;
